core/server/worker/src/main/java/alluxio/worker/block/allocator/MaxFreeAllocator.java

```diff
@@ -62,14 +62,22 @@ private StorageDirView allocateBlock(long sessionId, long blockSize,
 
     if (location.equals(BlockStoreLocation.anyTier())) {
       for (StorageTierView tierView : mManagerView.getTierViews()) {
-        candidateDirView = getCandidateDirInTier(tierView, blockSize);
+        candidateDirView = getCandidateDirInTier(tierView, blockSize,
+            BlockStoreLocation.ANY_MEDIUM);
```

Comment:
Should we be using the medium provided by `location`?

Reply:
in this case location has a mediumType of ANY_MEDIUM, that is why it is equal to `anyTier()`.

Reply:
Ah I see, that was not immediately obvious since tier and medium are orthogonal. The BlockStoreLocation is not straightforward anymore, could you add some documentation which includes some of the assumptions we make (for example that tier and medium are not concurrently set (I think this might be a valid use case?))